### PR TITLE
No more Jane Syntax for layouts

### PR DESCRIPTION
**Pull request by ncik-roberts**
Encode layouts-related things in the parsetree. In practice, this is unboxed constants, jkind annotations on type variables in various places, and kind abbreviations.

This diff should give you a warm, fuzzy feeling. Let me know any extent to which that is not the case.

Much of this diff is mechanical.

Some of the mechanical changes have the appearance of not being mechanical. Mostly, those are ones where a file has both "remove" and "add" diff: I move a code block from its old jane-syntax-style function back into the normal match.

Some non-mechanical things I'll call out:
  * Checking language extensions no longer happens for free. I add specific language extension checks for unboxed constants and jkind annotations.
  * I rework the parsetree for jkind annotations to have the usual `foo_desc`/`foo` split. This makes the ast iterating/mapping code more uniform with other constructs.
  * `printast.ml` and `pprintast.ml`, as usual, require some judgment. The tests for `pprintast.ml` give us some confidence I haven't screwed this up.
  * `ocaml/typing/typedecl.ml` contains some slightly more manual reworks. (I get to eliminate one of the disjuncts of an `Either`.)

### CODE REVIEW

File: ocaml/parsing/ast_iterator.ml [hunk 1]
```diff
@@ -246,31 +213,19 @@ module T = struct
 
   let iter_extension_constructor_kind sub = function
       Pext_decl(vars, ctl, cto )->
-        List.iter (iter_loc sub) vars;
+        List.iter (bound_var sub) vars;
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/parsing/ast_iterator.ml:216]: it is, and I'm resolving this along with your other comment

File: ocaml/parsing/depend.ml [hunk 2]
```diff
@@ -95,19 +95,16 @@ let handle_extension ext =
 
 (* CR layouts: Remember to add this when jkinds can have module
    prefixes. *)
-let add_jkind _bv (_jkind : Jane_syntax.Jkind.annotation) = ()
+let add_jkind _bv (_jkind : jkind_annotation) = ()
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/parsing/depend.ml:98]: Dammit! Dammit! (Sorry, I was taught the word "recur".) And, fixed.

File: ocaml/parsing/ast_mapper.ml [hunk 3]
```diff
@@ -181,27 +180,13 @@ module T = struct
...
-        let attrs = sub.attributes sub attrs in
-        let jtyp = sub.typ_jane_syntax sub jtyp in
-        Jane_syntax.Core_type.core_type_of jtyp ~loc ~attrs
-    end
-    | None ->
-    let attrs = sub.attributes sub attrs in
-    match desc with
-    | Ptyp_any -> any ~loc ~attrs ()
-    | Ptyp_var s -> var ~loc ~attrs s
+ *)
```
[the remark this answers is not in the document]
Comment (ncik-roberts) [reply ocaml/parsing/ast_mapper.ml:183]: yep, thanks

**Review by ccasin — APPROVED** [review]
Looks great, thanks!

**Pull request merged**